### PR TITLE
SONARJAVA-512 Unconditional jumps check

### PR DESCRIPTION
**Pull request by m-g-sonar**

### CODE REVIEW

File: java-checks/src/main/java/org/sonar/java/checks/UnconditionalJumpStatementCheck.java [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+      return;
+    }
+
+    if (tree.is(Tree.Kind.CONTINUE_STATEMENT) || !canExecuteMoreThanOnce(parent)) {
+      SyntaxToken jumpKeyword = jumpKeyword(tree);
+      reportIssue(jumpKeyword, String.format("Remove this \"%s\" statement or make it conditional.", jumpKeyword.text()));
+    }
+  }
+
+  private static boolean canExecuteMoreThanOnce(Tree loopTree) {
```
Comment (benzonico) [inline java-checks/src/main/java/org/sonar/java/checks/UnconditionalJumpStatementCheck.java]: I would name that method executeUnconditionnally to make things clearer on reporting side. and add a comment inside the method to explain that executing unconditionally means that we cannot find a path in the CFG that goes twice through this instruction.

File: java-checks/src/main/java/org/sonar/java/checks/UnconditionalJumpStatementCheck.java [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+    } else if (jumpStatement.is(Tree.Kind.RETURN_STATEMENT)) {
+      return ((ReturnStatementTree) jumpStatement).returnKeyword();
+    } else {
+      return ((ThrowStatementTree) jumpStatement).throwKeyword();
+    }
+  }
+
+  private static CFG getCFG(Tree loop) {
+    Tree currentTree = loop;
+    do {
```
Comment (benzonico) [inline java-checks/src/main/java/org/sonar/java/checks/UnconditionalJumpStatementCheck.java]: Wondering why you need to get back to the parents ? couldn't you consider the loop passed in tree as a list of statement and construct CFG solely for this one ? If i'm correct this should be sufficient.

File: java-checks/src/main/java/org/sonar/java/checks/UnconditionalJumpStatementCheck.java [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+
+  private static boolean canExecuteMoreThanOnce(Tree loopTree) {
+    CFG cfg = getCFG(loopTree);
+    return cfg.blocks().stream().anyMatch(block -> loopTree.equals(block.terminator()) && hasPredecessorInLoopBody(block.predecessors(), loopTree));
+  }
+
+  private static boolean hasPredecessorInLoopBody(Set<Block> predecessors, Tree loop) {
+    for (CFG.Block predecessor : predecessors) {
+      List<Tree> predecessorElements = predecessor.elements();
+      if (!predecessorElements.isEmpty()) {
```
Comment (benzonico) [inline java-checks/src/main/java/org/sonar/java/checks/UnconditionalJumpStatementCheck.java]: if else with a negation can be rewritten a bit more simpler

File: java-checks/src/main/java/org/sonar/java/checks/UnconditionalJumpStatementCheck.java (hunk 2 above)
Comment (m-g-sonar) [inline java-checks/src/main/java/org/sonar/java/checks/UnconditionalJumpStatementCheck.java]: Unfortunatly I can't. In some cases, you may decide to continue to a `label` which is defined outside the body of the loop. I added a test case which fails if I only take it. See test file [Line 108](https://github.com/SonarSource/sonar-java/pull/1306/files#diff-07fb5f5c2456e872e6fa9a2b8291bde7R108).

File: java-checks/src/main/java/org/sonar/java/checks/UnconditionalJumpStatementCheck.java [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+   * }
+   * </code>
+   */
+  private static boolean isEmptyConditionLoop(Tree loopTree) {
+    switch (loopTree.kind()) {
+      case FOR_STATEMENT:
+        ForStatementTree fst = (ForStatementTree) loopTree;
+        return fst.initializer().isEmpty() && fst.condition() == null && fst.update().isEmpty();
+      case WHILE_STATEMENT:
+        // 'while(false)' does not compile, unreachable code
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/checks/UnconditionalJumpStatementCheck.java]: MAJOR This block of commented-out lines of code should be removed. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACommentedOutCodeLine)

File: java-checks/src/main/java/org/sonar/java/checks/UnconditionalJumpStatementCheck.java [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+        return isTrue(((WhileStatementTree) loopTree).condition());
+      case DO_STATEMENT:
+        // only target true literal, which needs a break statement.
+        // For a 'do {...} while (false);', loop the jump is useless and issue should be raised.
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/checks/UnconditionalJumpStatementCheck.java]: MAJOR This block of commented-out lines of code should be removed. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACommentedOutCodeLine)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues
* ![MAJOR][MAJOR] 2 major

Watch the comments in this conversation to review them.

[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'

**Pull request merged**